File: dxbase/src/main/java/com/dxmovie/dxbase/activity/DXBaseActivity.java
package com.dxmovie.dxbase.activity;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.lifecycle.ViewModelProvider;

import com.dxmovie.dxbase.viewmodel.BaseViewModel;


/**
 * 封装VM的BaseActivity
 */
public abstract class DXBaseActivity<V extends ViewDataBinding, VM extends BaseViewModel> extends XBaseActivity {

    public V binding;
    public VM viewModel;
    private int viewModelId;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //初始化databinding和viewModel
        initViewDataBinding(savedInstanceState);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        //解绑dataBinding
        if (binding != null) {
            binding.unbind();
        }
    }


    public void initViewDataBinding(Bundle savedInstanceState) {
        binding = DataBindingUtil.setContentView(this, getContentViewLayout(savedInstanceState));
        viewModelId = getViewModelId();
        viewModel = getViewModel();
        //关联ViewModel
        binding.setVariable(viewModelId, viewModel);

    }

    /**
     * 获取页面的布局文件
     *
     * @param savedInstanceState
     * @return 布局的id
     */
    public abstract int getContentViewLayout(Bundle savedInstanceState);

    /**
     * 获取ViewModel的id
     *
     * @return 变量的id
     */
    public abstract int getViewModelId();

    /**
     * 初始化ViewModel
     *
     * @return
     */
    public abstract VM getViewModel();

    /**
     * 创建ViewModel
     *
     * @param cls
     * @param <T>
     * @return
     */
    public <T extends BaseViewModel> T createViewModel(Class<T> cls) {
        return new ViewModelProvider(this).get(cls);
    }

    @Override
    protected void onResume() {
        super.onResume();
    }
}

File: dxbase/src/main/java/com/dxmovie/dxbase/activity/DXBaseActivity.kt
package com.dxmovie.dxbase.activity

import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.lifecycle.ViewModelProvider
import com.dxmovie.dxbase.viewmodel.BaseViewModel
/**
 * 业务基类AC
 * MVVM 功能继承类
 */
abstract class DXBaseActivity<V : ViewDataBinding, VM : BaseViewModel> : XBaseActivity() {

    var binding: V? = null
    var vModel: VM? = null
    private var viewModelId = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //初始化databinding和viewModel
        initViewDataBinding(savedInstanceState)
    }

    override fun onDestroy() {
        super.onDestroy()
        //解绑dataBinding
        binding?.unbind()
    }

    open fun initViewDataBinding(savedInstanceState: Bundle?) {
        binding = DataBindingUtil.setContentView(this, getContentViewLayout(savedInstanceState))
        viewModelId = getViewModelId()
        vModel = getViewModel()
        //关联ViewModel
        binding!!.setVariable(viewModelId, vModel)
    }

    /**
     * 获取页面的布局文件
     *
     * @param savedInstanceState
     * @return 布局的id
     */
    abstract fun getContentViewLayout(savedInstanceState: Bundle?): Int

    /**
     * 获取ViewModel的id
     *
     * @return 变量的id
     */
    abstract fun getViewModelId(): Int

    /**
     * 初始化ViewModel
     *
     * @return
     */
    abstract fun getViewModel(): VM?

    /**
     * 创建ViewModel
     *
     * @param cls
     * @param <T>
     * @return
    </T> */
    open fun <T : BaseViewModel?> createViewModel(cls: Class<T>?): T {
        return ViewModelProvider(this)[cls!!]
    }

}